### PR TITLE
Add support for the management-Realm 

### PR DESCRIPTION
**Pull request by cmelchior**
Adds the low-level support for the Permission Realm:
- [x] Added permission classes
- [x] Added `getPermissionRealm` on `User`
- [x] Add Javadoc
- [x] Figure out how we get the URL to the management Realm
- [x] Enabled support for including Realm classes/modules inside the library
- [x] Replaced APT with the newly added AnnotationProcessor directive.

### CODE REVIEW

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 1]
```diff
@@ -354,6 +358,40 @@ public String getAccessToken() {
...
+            permissionRealmConfig = new SyncConfiguration.Builder(this, managementUrl)
+                    .schema(PermissionOffer.class, PermissionRequest.class, PermissionChange.class)
+                    .build();
+        }
+
+        return Realm.getInstance(permissionRealmConfig);
+    }
+
+    // Creates the URL to the permission Realm based on the authentication URL.
+    private static String getPermissionManagementRealmUrl(URL authUrl) {
```
Comment (mrackwitz) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: -> `getManagementRealmUrl`

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionOffer.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import java.util.UUID;
+
+import io.realm.RealmObject;
+import io.realm.annotations.PrimaryKey;
+import io.realm.annotations.Required;
+
+/**
+ * TODO
+ */
+public class PermissionOffer extends RealmObject {
```
Comment (mrackwitz) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionOffer.java]: Won't be initially available.

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionRequest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.UUID;
+
+import io.realm.RealmObject;
+import io.realm.annotations.PrimaryKey;
+import io.realm.annotations.Required;
+
+/**
+ * TODO
+ */
+public class PermissionRequest extends RealmObject {
```
Comment (mrackwitz) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionRequest.java]: Won't be initially available.

**Comment by cmelchior** [comment]
Ready for review @realm/java

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 4]
```diff
@@ -354,6 +357,39 @@ public String getAccessToken() {
...
+                    .build();
+        }
+
+        return Realm.getInstance(managementRealmConfig);
+    }
+
+    // Creates the URL to the permission Realm based on the authentication URL.
+    private static String getManagementRealmUrl(URL authUrl) {
+        String scheme = "realm";
+        if (authUrl.getProtocol().equals("https")) {
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: maybe `equalsIgnoreCase` for capital URL?

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Base fields
+    @PrimaryKey
+    @Required
+    private String id = UUID.randomUUID().toString();
+    @Required
+    private Date createdAt = new Date();
+    @Required
+    private Date updatedAt = new Date();
+    private Integer statusCode; // null=not processed, 0=success, >0=error
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java]: we could also use default value.-1 not processed, 0 success, >= 0 error?

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    private String statusMessage;
+
+    @Required
+    private String realmUrl;
+    @Required
+    private String userId;
+    private Boolean mayRead;
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java]: Ditto

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 7]
```diff
@@ -354,6 +357,39 @@ public String getAccessToken() {
         return (userToken != null) ? userToken.value() : null;
     }
 
+    /**
+     * Returns an instance of the Management Realm owned by the user.
+     *
```
Comment (kneth) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: `<p />`

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+import java.util.Date;
+import java.util.UUID;
+
+import io.realm.RealmObject;
+import io.realm.annotations.PrimaryKey;
+import io.realm.annotations.Required;
+
+/**
+ * This class is used for requesting changes to a Realms permissions.
```
Comment (kneth) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java]: "Realms" -> "Realm's"

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 9]
```diff
@@ -354,6 +357,39 @@ public String getAccessToken() {
...
+     */
+    public synchronized Realm getManagementRealm() {
+        if (managementRealmConfig == null) {
+            String managementUrl = getManagementRealmUrl(syncUser.getAuthenticationUrl());
+            managementRealmConfig = new SyncConfiguration.Builder(this, managementUrl)
+                    .modules(new PermissionModule())
+                    .build();
+        }
+
+        return Realm.getInstance(managementRealmConfig);
```
Comment (zaki50) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: We don't need to put `return Realm.getInstance(managementRealmConfig);` inside of the `synchronized`.

It's better to make a `synchronized` block and move `if (managementRealmConfig == null) {...}` into it.

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 10]
```diff
@@ -25,6 +25,7 @@
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.util.Arrays;
+import java.util.Collections;
```
Comment (dalinaum) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: It looks like it is added accidently. It is never used in `SyncConfiguration.java`.

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionModule.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package io.realm.permissions;
```
Comment (dalinaum) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionModule.java]: Does it need a license header too?

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 12]
```diff
@@ -66,7 +66,7 @@
     public static final int KEY_LENGTH = 64;
 
     private static final Object DEFAULT_MODULE;
-    private static final RealmProxyMediator DEFAULT_MODULE_MEDIATOR;
+    static final RealmProxyMediator DEFAULT_MODULE_MEDIATOR;
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: It was used by the `schema()` in SyncConfiguration, but that was removed again, so putting back `private`

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionModule.java (hunk 11 above)
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionModule.java]: Done

File: realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+    public PermissionChange() {
+        // Default constructor required by Realm
+    }
+
+    /**
+     * Construct a Permission Change Object.
+     *
+     * @param realmUrl Realm to change permissions for.
+     * @param userId User or users to effect.
```
Comment (mrackwitz) [inline realm/realm-library/src/objectServer/java/io/realm/permissions/PermissionChange.java]: This could mention wildcards and how they work.

**Pull request merged**